compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -1038,6 +1064,23 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] {
     else None
   }
 
+  /** Check whether we can compare the given set of atoms with another to determine
+   *  a subtype test between OrTypes. There is one situation where this is not
+   *  the case, which has to do with SkolemTypes. TreeChecker sometimes expects two
+   *  types to be equal that have different skolems. To account for this, we identify
+   *  two different skolems in all phases `p`, where `p.isTyper` is false.
+   *  But in that case comparing two sets of atoms that contain skolems
+   *  for equality would give the wrong result, so we should not use the sets
+   *  for comparisons.
```

Comment:
Isn't not using the sets a potential performance problem ? Have you considered having both `OrType#atoms` and `OrType#atomsUnifySkolems` instead, where the latter would replace all skolems referring to the same type by a single instance ?

Reply:
I believe skolems are sufficiently rare. If it turns out that we get a performance problem, what you suggest could be a good alternative.